Parallelism tutorial (#3886)
oxcaml/oxcaml: jane/doc/parallelism-tutorial.md

Reviewer: I think folks reading this may also want to be able to observe the speedup. Though I don't think you should spend a lot of time explaining benchmarking or anything, I think it would be a valuable addition to show an actual executable that can be run to get timing measurements -- and then to show that parallelism can speed things up.

Reviewer: Might be nice to have a little summary at the end. This would include tips like "look at the Parallel library" or "benchmark after parallelizing". It could also maybe have all your Rules stated together in a little table.

Author: Did end up making a table. A conclusion would be good but can be added later.

# Parallelism Tutorial

# Introduction

# Fork/join parallelism

Suppose we're working with binary trees, and we want to take an average over all
the values in the tree. Here's a basic implementation:

<a id="code-average"></a>
```ocaml
module Tree = struct
  type 'a t =
    | Leaf of 'a
    | Node of 'a t * 'a t
end

> let average (tree : float Tree.t) =

Author: Adding a link to the paper (seems like the canonical documentation at the moment)

> let average (tree : float Tree.t) =

Reviewer: What paper? This is about labeled tuples, which I don't think are in a paper...

> let average (tree : float Tree.t) =

Author: Well, there's at least this paper (workshop paper I think?): https://tyconmismatch.com/papers/ml2024_labeled_tuples.pdf

  let rec total tree =
    match tree with
    | Tree.Leaf x -> ~total:x, ~count:1
    | Tree.Node (l, r) ->
      let ~total:total_l, ~count:count_l = total l in
      let ~total:total_r, ~count:count_r = total r in
      ~total:(total_l +. total_r), ~count:(count_l + count_r)
  in
  let ~total, ~count = total tree in
  total /. (count |> Float.of_int)
;;
```

Now suppose we want to take advantage of multicore OCaml and get those sweet,
sweet parallel speedups. This is an obvious candidate for _fork/join

> parallelism,_ as provided by the `parallel` library’s `fork_join*` functions.

Reviewer: As on Slack: sadly, no. But we can link to the mli file.

          Parallel.fork_join2
            par
            (fun par -> total par l)
            (fun par -> total par r)
        in
        ~total:(total_l +. total_r), ~count:(count_l + count_r)
    in
    let ~total, ~count = total par tree in
    total /. (count |> Float.of_int)
  ;;
```

Note that we don't have to worry about unbalanced trees: the work-stealing
algorithm dynamically adapts whenever tasks are unevenly distributed among cores.

We're not limited to working with simple `float`s, of course. Suppose we add a
submodule `Thing` earlier in the file: **[Would be nice to think of something
better than `Thing` here.]**

<a id="listing-thing-impl"></a>
```ocaml
module Thing = struct
  module Mood = struct
    type t =
      | Happy
      | Neutral
      | Sad
  end

  type t =
    { price : float
    ; mutable mood : Mood.t
    }

  let create ~price ~mood = { price; mood }
  let price { price; _ } = price
  let mood { mood; _ } = mood
end
```

All we have to do to sum over the prices in a `Thing.t Tree.t` is change the
`Tree.Leaf` case:

```ocaml
  let average_things_par (par : Parallel.t) tree =
    let rec total par tree =
      match tree with

>       | Tree.Leaf x -> ~total:(Thing.price x), ~count:1 (* <== new code *)

Author: Trying it as a diff with two lines of context. Seems good

```ocaml
val price : t -> float @@ portable
```

But the compiler isn't satisfied. Now it complains about the `l` in
`total par l`:

```
This value is contended but expected to be uncontended.
```

Now the solution is much less obvious, but as we'll see, it again comes down to
missing information on `price`:

```ocaml
val price : t @ contended -> float @@ portable
```

Once you're familiar with the `portable` and `contended` modes, you'll see that
this can be read as saying:

> The `price` function is safe to call from any domain, and it won't produce a
> data race even if its argument is mutated in parallel.

That's enough to let `average_par_things` go through: from there, the compiler
infers the safety properties that `Parallel.fork_join2` demands.

Adding `portable` and `contended` to your .mli files can take some work, but
over the next few sections we'll cover what exactly the modes mean, what the
most important rules for using them are, and some workarounds and shortcuts. In
return, these two modes suffice to run even large amounts of code in fork/join
style without fear of data races.

## How `portable` and `contended` work

> Firstly, `portable` and `contended` are _modes_. Like a type, a mode describes

Author: Yeah, I've tried to keep this section lean (emphasis on “tried”) in part for this reason. I do think there's value in going over this stuff even if there's overlap with other documentation.

not permissible_ to do so. Unlike rule 2, there's no one place where rule 1 is
enforced—really, it's an invariant of the whole system (language, runtime, and
core libraries)—but let's see one example of trying to break it:

```ocaml
let beat_the_system par =
  let t = { price = 42.0; mood = Neutral }
  let (), () =
    Parallel.fork_join2 par (fun _par -> cheer_up t) (fun _par -> bum_out t)
  in
  ()
```

```
Error: This value is contended but expected to be uncontended.
```

When we cover the `portable` mode [next], we'll be able to explain precisely
what's going on here, but for the moment suffice it to say that the type of
`fork_join2` forces the two tasks to obey rule 1.

[next]: #the-portable-and-nonportable-modes

You may worry that the different modes on `price` and `mood` prevent writing
sensible code like

```ocaml
let price_and_mood (t @ uncontended) = price t, mood t
```

Can we really treat `t` as both `contended` (to call `price`) and `uncontended`
(to call `mood`)? Actually, we can:

<a id="rule-contended-submode"></a>
> **Rule 3.** An `uncontended` value may be used as though it is `contended`.

> This is entirely consistent with rules 1 and 2: having an `uncontended` value

Author: Hmm. Really there are two audiences, though, and for the Jane Street audience the chances are much higher that they'll have seen it, and it seems making the connection. Maybe if I move the parenthetical sentence here to a footnote? That would more strongly suggest that it's not important to understand. Or I could keep it in place so that people see it but phrase it as “If you've worked with `local`, you may have seen that it works the same way.”

> This is entirely consistent with rules 1 and 2: having an `uncontended` value

Author: Downgrading the passage here to a footnote. I doubt writing for the non-JS audience would annoy the JS audience, I'd just hate to lose the opportunity to leverage something the reader might already know

> This is entirely consistent with rules 1 and 2: having an `uncontended` value

Reviewer: I think a footnote makes sense. And actually I wouldn't assume the JS audience knows about locality either. Most devs I speak with don't. They've vaguely heard of it, but have no idea what it means.

of every constructor of a `contended` variant is `contended`. And of course this
applies recursively, so the components of _those_ components are also
`contended`. Accordingly, We say that the `contended` mode is _deep._ It's easy
to see what goes wrong if we let anyone treat, say, a field as `uncontended`
when its record is `contended`:

<a id="code-cheer_up_sneakily"></a>
```ocaml
type t_in_a_trenchcoat = {
  inner_t : t;
}

let cheer_up_sneakily (t_in_a_trenchcoat @ contended) =
  let t @ uncontended = t_in_a_trenchcoat.inner_t (* error: rule 4 *) in
  cheer_up t (* cue the ominous music again *)
```

If not for rule 4, we could also write `bum_out_sneakily` and then call
`cheer_up_sneakily` and `bum_out_sneakily` on the same argument in parallel,
causing the same data race we've been trying to avoid all along.

In fact, rule 4 is pretty well required to make [rule 1] work at all. The
interested reader is invited to work out why the text of rule 1 demands that,
for instance, we say that `t_in_a_trenchcoat.inner_t` is `contended` when
`t_in_a_trenchcoat` is `contended`.

[rule 1]: #rule-contended-parallel

> #### Modalities

Author: Sorry, not sure what general treatment you're referring to? I tend to agree, we can probably lose it. The one thing that makes me think otherwise is that I explain at some length why, say, a `contended` thing can _only ever_ contain `contended` things, and then it's valuable to say that `uncontended` things only contain `uncontended` things _by default._ But we probably don't need the whole song and dance, especially if I can link somewhere else for the details.

> #### Modalities

Author: Removed both “Modalities” sections for now.

[Rule 4] says the `contended` mode is deep—that `contended` values can't
contain `uncontended` values (unless we demote them to `contended` by [rule 3]).
In contrast, there's such requirement for `uncontended` values: they can contain
either `contended` or `uncontended` values, so long as we keep track of which is
which. For convenience, `uncontended` is “deep by default,” and we use a
notation called a _modality_ to make a field `contended` instead:

[Rule 4]: #rule-contended-deep
[rule 3]: #rule-contended-submode

```ocaml
module State = struct
  type t = {
    mutable most_expensive_thing : Thing.t @@ contended

>     mutable my_favorite_thing : Thing.t

Author: Urk, I think I must have had this realization in my test code but failed to transfer that knowledge to the tutorial. (My test code has the record but neither field is `mutable`.)

```ocaml
@@ portable

type t

val create : unit -> t
val combine : t -> t -> t
val do_something_involving_shared_state : t -> unit @@ nonportable
```

## Mode crossing

In discussing why [`portable` is deep], we mentioned that very few types care
about `portable`, since the values of most types aren't functions and don't
contain functions. Similarly, if a type `t` doesn't have any `mutable` fields or
mutable arrays, having a `contended` or `uncontended` `t` makes no difference.

[`portable` is deep]: #rule-portable-deep

Fortunately, the compiler can be made aware that a type “doesn't care” about
`portable` or `contended`:

```ocaml
type float_and_int : immutable_data = float * int
```

The `immutable_data` here is called a _kind._ Just as the value `(1.2, 3)` has
the type `float_and_int`, in turn the type `float_and_int` has the kind
`immutable_data`. Also like types, kinds are inferred where possible, so in
fact the `: immutable_data` in this case is redundant. However, in an .mli
they are often useful.

The three most important kinds for data-race freedom are `immutable_data`,
`mutable_data`, and `value`.

| Kind | Requirements | Crosses |
| ---- | ------------ | ------- |

> | `immutable_data` | no functions or mutable fields, deeply | portability, contention |

Author: Ah, good point. I've added a blurb about not being exhaustive—we could tease the names of other axes here, but the nice thing about the table as is is that the reader might hope to explain the whole table to themselves. (Also if I make the table exhaustive it will become outdated every month or two.)

Am I right that `value` truly crosses nothing? If there's something it doesn't cross, than I should write “neither” rather than “none”.

> | `immutable_data` | no functions or mutable fields, deeply | portability, contention |

Reviewer: Today, `value` crosses nothing. But it's still `mod non_null separable` (where the `separable` part comes with the next compiler roll). But those things aren't modes, so your statement that `value` crosses nothing is still valid.

_carry the `mutable` keyword._ It's right there in [rule 2 of `contended`]:
no “reading or writing to a `mutable` field.”

[rule 2 of `contended`]: #rule-contended-mutable
-->

We've seen that `mutable` fields can cause quite some trouble: they require
`uncontended` access, which can make functions not `portable`, and generally
they very much get in the way. In some cases, however, you _may_ (with
significant caveats) be able to swap out `mutable` for `Atomic.t`. Let's do
that with our `thing.ml`:

```ocaml
module Thing = struct
  module Mood = struct
    type t =
      | Happy
      | Neutral
      | Sad
  end

  type t : immutable_data =
    { price : float
    ; mood : Mood.t Atomic.t
    }

  let create ~price ~mood = { price; mood = Atomic.make mood }
  let price { price; _ } = price
  let mood { mood; _ } = Atomic.get mood
  let cheer_up { mood; _ } = Atomic.set mood Happy
  let bum_out { mood; _ } = Atomic.set mood Sad
end
```

As you can see, there's a bit of syntactic overhead, but in return we get to
access `mood` even if a `Thing.t` is `contended`, and in fact we can mark `t` as
`immutable_data` so that it ignores `contended` and `uncontended` altogether

> (that is, it [crosses contention]). Do go over the documentation for the `Atomic`

Author: Urk. Well, I can link to the stdlib documentation, which doesn't have the right modes and isn't quite the same as Base/Core. Leaving this dangling for the moment

> (that is, it [crosses contention]). Do go over the documentation for the `Atomic`

Reviewer: Just link to the Base/Core mli file, I think.

the total and count from our fork/join tasks, we can keep the running total and
count in atomics:

```ocaml
  let average_par_running (par : Parallel.t) tree =
    let total = Atomic.make 0.0 in
    let count = Atomic.make 0 in
    let rec go par tree =
      match tree with
      | Tree.Leaf x ->
        Atomic.update total ~pure_f:(fun total -> total +. Thing.price x);
        Atomic.incr count
      | Tree.Node (l, r) ->
        let (), () =
          Parallel.fork_join2 par (fun par -> go par l) (fun par -> go par r)
        in
        ()
    in
    go par tree;
    Atomic.get total /. (Atomic.get count |> Float.of_int)
```

If we kept `total` and `count` in `ref`s, then `go` would not be able to access
them. (Exercise: What rules combine to stop us? Remember, a `ref` is just a
record whose only field is `mutable`. Don't rely on rule 1 of `contended` or
rule 1 of `portable` for your answer—those only tell us that we _must be
stopped,_ not what actually stops us.)

Now for the caveats: Firstly, there are performance penalties, since an
`Atomic.t` is a pointer and atomic operations are more expensive. Secondly,

> `Atomic.t` frees us from concerns about data races, but it absolutely does not

Author: Yeah, this is def one of the things that needs to be in the introduction

prevent race _conditions_. There are few guarantees about the order in which the
updates to `total` and `count` occur: we know only that they all happen before
the outermost `fork_join2` returns.

Of course, in this case what saves us is that it doesn't matter what order the
updates happen in: addition is commutative. However, now that we've made `mood`
atomic as well, suppose we wanted to take the average of all the nodes whose
`mood` is `Happy`? Nothing in the system stops us now: if someone is calling
`bum_out` on the entire tree in parallel, then our average will reflect an
unpredictable number of those changes. The only way to stop _that_ is to use
something more sophisticated like a lock over the whole tree, which grants a
function `uncontended` access while the lock is held. That's what the capsule

> API is for, but it's beyond the scope of this tutorial.

Reviewer 2: I wouldn't say memory corruption exactly since the memory model already prevents undefined behavior. Maybe state corruption?

> API is for, but it's beyond the scope of this tutorial.

Author: Hmm. I'm learning that my belief in what DRF prevents was imprecise. I'll ask Stephen when he's in tomorrow

> API is for, but it's beyond the scope of this tutorial.

Reviewer: Good point about memory corruption. I had been neglecting that fact about the memory model. Maybe our type system only prevents memory corruption in the presence of unboxed products like internal pointers? That is, a data race on a mutable unboxed product -- where one element in the product determines the validity of another -- can cause arbitrary memory corruption. But OCaml does not have such products natively. We're about to add them -- in the next week or two, according to @ccasin. So once we have such things, despite the guarantees the memory model gives us, then we'll have the possibility for memory corruption. And our type system prevents it. We both giveth and taketh away.

Neglecting this corner case (around dependent unboxed products), I'm honestly not sure what we gain with DRF.

> API is for, but it's beyond the scope of this tutorial.

Author: Added a whole section to the beginning about this, having talked with Stephen. The headline is we get sequential consistency (but of course I have to explain what that is).

`Parallel.Sequence.of_iarray`, but now `Parallel.Sequence.fold'` combines
the `map` and `reduce` operations:

```ocaml
let average_par (par : Parallel.t) tree =
  let rec (total @ portable) par tree =
    match tree with
    | Tree.Leaf x -> ~total:(Thing.price x), ~count:1
    | Tree.Node arr ->
      let seq = Parallel.Sequence.of_iarray arr in
      Parallel.Sequence.fold'
        par
        seq
        ~f:(fun par subtree -> total par subtree)
        ~init:(~total:0.0, ~count:0)
        ~combine:(fun _par (~total, ~count) (~total:total2, ~count:count2) ->
          ~total:(total +. total2), ~count:(count + count2))
      [@nontail]
  in
  let ~total, ~count = total par tree in
  total /. (count |> Float.of_int)
```

(We use `fold'` rather than `fold` to get the version that passes `par` down
into our `f` and `combine` functions.)

Another approach would be to convert the entire tree into a
`Parallel.Sequence.t` at once rather than rely on nested parallelism. This is
possible using `Parallel.Sequence.unfold`, though it does take more work,
requiring a definition of an unfolding state with an operation that splits it in
half.

> **[I made a solution for this but unless there's something I missed that would

Reviewer: Hard for me to see exactly where this is going, so I don't have much of an opinion here. I do like the idea of generalizing the initial example to something a little harder.

functions. For example, a very expensive way to add four integers (I promise the
other examples are more substantial) would be this:

<a id="code-add4"></a>
```ocaml
let add4 (par : Parallel.t) a b c d =
  let a_plus_b, c_plus_d =
    Parallel.fork_join2 par (fun _par -> a + b) (fun _par -> c + d)
  in
  a_plus_b + c_plus_d
```

The call to `Parallel.fork_join2` will schedule the calculations of `a + b` and
`c + d` as independent *tasks*, returning both once they're both done, and then
`add4` finishes by adding the results from the tasks. The `par` argument
parameterizes `fork_join2` (and, in turn, `add4`) by a particular implementation
of parallelism. It is also passed to the tasks so that they can spawn sub-tasks.

To run `add4`, we need to get our hands on a *scheduler*. Each scheduler is
provided by a library and determines the policy by which tasks get doled out and
run in domains. For this tutorial, we'll use `parallel_scheduler_work_stealing`,
which implements the popular [work-stealing] strategy.

[work-stealing]: https://en.wikipedia.org/wiki/Work_stealing

```ocaml
let test_add4 par = add4 par 1 10 100 1000

> let%expect_test "add4 in parallel" =

Author: Ah, good point. Probably fine to just use a top-level printf here and put the result of execution in the next text box?

We can use `fork_join2` to parallelize `average`:

<a id="listing-average_par"></a>
```ocaml
let average_par (par : Parallel.t) tree =
  let rec total par tree : total:float * count:int =
    match tree with
    | Tree.Leaf x -> ~total:x, ~count:1
    | Tree.Node (l, r) ->
      let (~total:total_l, ~count:count_l), (~total:total_r, ~count:count_r) =
        Parallel.fork_join2
          par
          (fun par -> total par l)
          (fun par -> total par r)
      in
      ~total:(total_l +. total_r), ~count:(count_l + count_r)
  in
  let ~total, ~count = total par tree in
  total /. (count |> Float.of_int)
;;
```

Note that we don't have to worry about unbalanced trees: the work-stealing
algorithm dynamically adapts whenever tasks are unevenly distributed among cores.

We're not limited to working with simple `float`s, of course. Suppose we add a
module `Thing` earlier in the file: **[Would be nice to think of something
better than `Thing` here.]**

<a id="listing-thing-impl"></a>
```ocaml
module Thing = struct
  module Mood = struct
    type t =
      | Happy
      | Neutral
      | Sad
  end

  type t =
    { price : float

>     ; mutable mood : Mood.t

Author: Ah, thanks—we seem to have that warning turned off

    }

  let create ~price ~mood = { price; mood }
  let price { price; _ } = price
  let mood { mood; _ } = mood
end
```

All we have to do to sum over the prices in a `Thing.t Tree.t` is change the
`Tree.Leaf` case:

```diff
  let rec total par tree : total:float * count:int =
    match tree with
-   | Tree.Leaf x -> ~total:x, ~count:1
+   | Tree.Leaf x -> ~total:(Thing.price x), ~count:1
    | Tree.Node (l, r) ->
      let (~total:total_l, ~count:count_l), (~total:total_r, ~count:count_r) =
```

So far, so good. But something annoying happens if we introduce an abstraction
barrier. Let's move `Thing` into its own file. `thing.mli` is very simple:

```ocaml
type t

module Mood : sig
  type t =
    | Happy
    | Neutral
    | Sad
end

val create : price:float -> mood:Mood.t -> t
val price : t -> float
val mood : t -> Mood.t
```

But now we get an error from the compiler:

```
The value Thing.price is nonportable, so cannot be used inside a function that

> is portable.

Author: Whoops, yeah I lost track that I'd already added a `@ portable` in my test code

```

```
Error: This value is contended but expected to be shared or uncontended.
```

This is dangerous for the same reason `cheer_up` was: someone else could be
running `bum_out` in parallel, producing a data race.[^shared]

[^shared]: You may have noticed the `shared` in the error message here. The
`shared` mode lies in between `contended` and `uncontended` in that it allows
reading but not writing mutable fields. It's less common than the others but it
comes in handy for things like read/write locks.

Adding `uncontended` signals to the compiler that a data race is possible: if
`cheer_up` and `bum_out` can be called with the same argument in parallel, we
have a race. Enter [rule 1](#rule-contended-parallel), which simply says _it is
not permissible_ to do so. Unlike rule 2, there's no one place where rule 1 is
enforced—really, it's an invariant of the whole system (language, runtime, and
core libraries)—but let's see one example of trying to break it:

```ocaml
let beat_the_system par =
  let t = { price = 42.0; mood = Neutral }
  let (), () =
    Parallel.fork_join2 par (fun _par -> cheer_up t) (fun _par -> bum_out t)
  in
  ()
```

```

> Error: This value is contended but expected to be uncontended.

Author: Oh! I never did say to add `@@ portable` to those functions, so that makes sense.

# Simple Parallelism Without Data Races

# Introduction

OCaml 5 unleashed parallel programming for the OCaml ecosystem. As is often the
case when something is unleashed, this has quite a bit of potential to cause
chaos. **[Need to quit fussing with this so I'm leaving it as is for the moment,
but am I saying we're putting the leash back on? Leashes are good, actually? Or

> maybe we're putting up a fence so the the multicore dog can run around without

Author: Decided in the end to keep the line but not extend the metaphor.

data races biting people??]**

OxCaml aims to provide an infrastructure for multicore OCaml programming that
exploits the full power of parallelism while _guaranteeing_ freedom from an
especially pernicious threat, that of the _data race._ That infrastructure
includes many parts, from common primitives like atomics and locks to newer
features like modes and capsules. Fortunately, not every parallel program will
need all of this machinery, and following this tutorial, you'll see how far you
can get with just two modes and the `Parallel.fork_join2` function.

## What is a data race?

A data race is a race condition in which two accesses to the same memory
location conflict. Specifically, there are four elements of the crime: **[[Oops,

> I forgot to define “domain” first]]**

Reviewer: Yes, maybe write a little section before this one entitle "A very brief introduction to multicore OCaml" or something

The one thing we want _not_ to happen is for domain B to see that
`!initialised` is `true` and yet find that `price_of_gold` is still stuck at
its initial value of `0.0`, thus causing us to `buy_lots_of_gold ()` no matter
the price. Our intuitive method reassures us that this cannot happen: domain A
doesn't set `initialised` to `true` until it's already done setting
`price_of_gold` to something sensible.

Unfortunately, our intuitive reasoning is wrong: this code very much can end up
calling `buy_lots_of_gold ()` despite `price_of_gold` being high. Since the
assignments to `price_of_gold` and `initialised` happen in the same domain and
neither is a fancy [atomic] operation, they can be reordered so that
`initialised` is set first, _while `price_of_gold` is still zero,_ leading
domain B to make an expensive mistake.

[atomic]: #atomics

It turns out our intuition assumes what the literature calls
_sequential consistency:_ essentially, the principle that things happen in order
they're written, and parallelism just means we're not sure which domain will do
its next thing. This principle is so essential that it's hard to imagine

> reasoning about a multicore program without it.

Reviewer: ```suggestion
reasoning about a multicore program without it. And yet, when a data race occurs,
sequential consistency goes out the window: we can observe mutations (and other
side effects) *completely out of order*.
```

Or something. I would love to incorporate the vision in my head of an oxcaml (the animal) wrestling the dog-prince of chaos to the ground and putting a leash on it, but I don't know how to do so. Maybe AI can help? Or maybe James Somers can, if he enjoys a little creative writing.

Author: Hmm. It's actually not quite that bad, though? The data races are, after all, bounded in time and space. Maybe I don't need to split those hairs for this, though?

Of course, what's happened here is two data races: each of `price_of_gold` and
`initialised` is being modified by domain A and read by domain B in parallel.
So the moral of the story is that data races violate sequential consistency.

The good news is that the inverse is also true: any program without data races

> is sequentially consistent. Data-race freedom gives us back the power to reason
> intuitively about our code, no matter how buggy it might get. In our example,

Reviewer: ```suggestion
is sequentially consistent. *Data-race freedom gives us back the power to reason
intuitively about our code, no matter how buggy it might get.* In our example,
```

that means that once we fix the data races (easy to do with `Atomic.t`), we know

> that fixes our bug: sequential consistency means that it's impossible for domain

Reviewer: ```suggestion
our bug is gone: sequential consistency means that it's impossible for domain
```

B to observe a state where `initialised` is `true` and `price_of_gold` is
`0.0`.[^price-zero]

[^price-zero]: Assuming, of course, `calculate_price_of_gold ()` doesn't
return `0.0`. The compiler can't catch everything.

> The even _better_ news is that, in code using OxCaml's extensions, the data

Reviewer: Great para.

races in `price_of_gold` and `initialised` would _never have compiled to begin
with._ There's a tradeoff, of course: just as OCaml's type safety means having
to understand the type system, the system for data-race freedom has a learning
curve. This tutorial aims to get you moving with a practical framework that
hopes to achieve significant parallel speedups without much fuss. You'll still
need to convince the compiler that your code has no data races, but we'll walk
through the rules and even some [shortcuts](#amenities) for common special
cases.

# Fork/join parallelism

> Any problem that can be neatly subdivided is a good candidate for

Reviewer: "subdivided into parallel pieces"? But that makes the sentence a little circular. I mean, *any* problem can be subdivided. So I think this isn't exactly what you mean (but fail to have a better phrasing at the moment).

Author: Yeah, I struggled a bit too. The word _neatly_ is sort of standing in for the idea that the pieces have to be separate but it's not very precise.

_fork/join parallelism,_ as provided by the `parallel` library’s `fork_join*`
functions. For example, a very expensive way to add four integers (I promise the

> other examples are more substantial) would be this:

Reviewer: I love the parenthetical. But it also begs the question: who is "I"? Are we planning on putting your name at the top of this? Perhaps we should; I see no reason not to. If not, though, maybe switch to "we".

This creates a work-stealing scheduler, along with a _monitor_ to manage
exceptions. Then it tells the scheduler to run the `test_add4` function before
shutting down the scheduler. (Naturally, a real program will want to keep the
monitor and scheduler around longer!)

You can also use the `parallel` library's own `Parallel.Scheduler.Sequential`,
which simply runs everything on the primary domain. This is handy for testing or
debugging when you want to eliminate nondeterminism. To do so, simply replace
`Parallel_scheduler_work_stealing` with `Parallel.Scheduler.Sequential`.

Now for something more substantial. Suppose we're working with binary trees, and
we want to take an average over all the values in the tree. Here's a basic
implementation:

<a id="code-average"></a>
```ocaml
module Tree = struct
  type 'a t =
    | Leaf of 'a
    | Node of 'a t * 'a t
end

let average (tree : float Tree.t) =

>   let rec total tree : total:float * count:int =

Reviewer: ```suggestion
  let rec total tree : (total:float * count:int) =
```

The parens are redundant, of course, but given the other `:` on the line, I think it makes the code easier to read.

    match tree with
    | Tree.Leaf x -> ~total:x, ~count:1
    | Tree.Node (l, r) ->
      let ~total:total_l, ~count:count_l = total l in
      let ~total:total_r, ~count:count_r = total r in
      ~total:(total_l +. total_r), ~count:(count_l + count_r)
  in
  let ~total, ~count = total tree in
  total /. (count |> Float.of_int)
;;
```

(Note that we've made use of the new [labeled tuples] feature for the values

> returned by `total`.)

Reviewer: I would move this note above the code. Otherwise, readers' OCaml parsers will fail and they may not make it here.

> [labeled tuples]: https://tyconmismatch.com/papers/ml2024_labeled_tuples.pdf

Reviewer: I think it's better to point to the documentation at oxcaml.org.

All we have to do to sum over the prices in a `Thing.t Tree.t` is change the
`Tree.Leaf` case:

```diff

>   let rec total par tree : total:float * count:int =

Reviewer: Did you mean the change the name of the function too?

Author: I've changed the references to keep using the old name.

[rule 1]: #rule-contended-parallel

Now, you may be surprised to see that projecting a field from a `contended`
record gives a `contended` value. After all, we have the function:

```ocaml
let price t = t.price
```

and we stated that its type is:

```ocaml
val price : t @ contended -> float @@ portable
```

Shouldn't the returned value be `contended`? In other circumstances it would: if
you try changing the `float` in the definition of `t` to a `float ref` then
you'll find you indeed have to add `contended` to the return type of `price`.
We'll cover the details in the section on [mode crossing], but the short answer

> is that since `float`s are immutable, no data race can arise from them, so the

Reviewer: ```suggestion
is that since `float`s are deeply immutable, no data race can arise from them, so the
```

compiler simply ignores `contended` and `uncontended` on `float`s (and you can

> make use of this for your own immutable data as well).

Reviewer: ```suggestion
make use of this for your own deeply immutable data as well).
```

<a id="rule-portable-deep"></a>
> **Rule 4.** Every component of a `portable` value must be `portable`.

You may have been wondering why rules 1 and 2a restrict all values, when really
it's only functions that we care about being `portable` or
not.[^other-code-types] This is the reason: many values can _contain_ functions,
and if it weren't for rule 4, we could use, say, a record to smuggle the
function around (in other words, the record could be a function [in a
trenchcoat]). So we have to forbid _all accesses_ of `nonportable` values from
other domains, not just function calls. The good news is that many types (in

> fact, most types) obviously _can't_ cause a problem—you'll never make an `(int *

Reviewer: Wrong kind of dash?

string option) list` that contains _any_ function, much less one that will cause

> a data race—and if the compiler is aware that a type `t` can't have a function

Reviewer: Other end of "wrong kind of dash?"

they are often useful.

The three most important kinds for data-race freedom are `immutable_data`,
`mutable_data`, and `value`. Some kinds constrain what types they describe, and
in return types with those kinds get to cross certain modes. In summary (this
table isn't nearly exhaustive—in particular, there are many modes we haven't

> covered, and `immutable_data` and `mutable_data` cross some of them as well):

Reviewer: Link to the now-extant kinds documentation somewhere around here.

(We have to write this slightly awkwardly to specify the returned mode of a
function.)

In fact, we've already seen mode crossing in action. We mentioned before when
talking about [rule 4 of `contended`](#rule-contended-deep) that when
`Thing.price` projects the `price` field out of a `contended` `Thing.t`, the
only reason it gets to return an `uncontended` `float` is that `float` crosses

> contention (which is to say, it has kind `immutable_data`).

Reviewer: Groan. Very sadly, `float : immutable_data` no longer holds. The problem is that `immutable_data` expands to `value mod ... non_float`. So, yeah, not `float`.

Author: Thought about hand-waving this a bit but I think it's easiest just to delete the parenthetical. There's already the blurb after the table that introduces the idea that some special types don't have these exact kinds but do cross those modes.
